### PR TITLE
video_core/texture_cache: Rewrite the texture cache

### LINKED ISSUE #3717: Project DIVA Mega Mix - Vulkan crashing after #3542
yuzu Mainline 206

GeForce RTX 2080
Ryzen R9 3950X

Vulkan crashes during boot-up starting from [Mainline 206](https://github.com/yuzu-emu/yuzu-mainline/releases/tag/mainline-0-206) and above. This is caused by #3542.

[yuzu_log.zip](https://github.com/yuzu-emu/yuzu/files/4497533/yuzu_log.zip)
[callstack.zip](https://github.com/yuzu-emu/yuzu/files/4497552/callstack.zip)

### LINKED ISSUE #3963: Super Mario Odyssey - line visible in shading on OpenGL
There is a line in the middle of the shading when using OpenGL.

yuzu mainline 258

GeForce RTX 2080
Ryzen R9 3950X

Emulator:

Console:

Vulkan does not have this issue.

Log:
[yuzu_log.zip](https://github.com/yuzu-emu/yuzu/files/4647941/yuzu_log.zip)

### LINKED ISSUE #4750: Pokémon Mystery Dungeon: Rescue Team DX - Graphics issue
switch:
2020100211225400-A5A477B869AFD225B3DAF74249D70FF2
yuzu:

01003d200baa2000_2020-10-02_12-27-36-592

there is minor graphic glitch when selecting pokemon 
my hardware is i7 8700k 16gb ram 2080s driver version 456.55
here is the log:
[mystery dungeun.txt](https://github.com/yuzu-emu/yuzu/files/5316004/mystery.dungeun.txt)

### LINKED ISSUE #4968: Animal Crossing New Horizon Update 1.6.0, black fog even with the mod Disable DOF on

### LINKED ISSUE #4979: Pokémon Sword graphical issues unique to EA
### **_You hurt my boy. How could you hurt my boy._**
Ahem, so I've got some graphical glitches on EA 1146 that does not happen on mainline. Comparison screenshots are attached. I blame the cache rewrite but that's just me 👀

The hardware in my pc that may or may not matter:
AMD Ryzen 5 3600 & AMD Radeon RX 580

In both mainline and EA I am using vulkan, async, and multicore (and docked if it matters)

### On to the screenshots
A log for this run: [yuzu_mainline.txt](https://github.com/yuzu-emu/yuzu/files/5587233/yuzu_mainline.txt)
As you can see in this screenshot. Everything is all good and nice.
yuzu_XDe96LlpQW

Log for this run: [yuzu_ea.txt](https://github.com/yuzu-emu/yuzu/files/5587235/yuzu_ea.txt)
In this one my boy is very sick. He needs some (medical) help
yuzu_ySZI1QZGLL

A video showing the bug in all it's terrifying glory https://youtu.be/jIOFtlOWRPc

### PR DESCRIPTION
**Pull request by ReinUsesLisp**
The current texture cache has several points that hurt maintainability and performance. It’s easy to break unrelated parts of the cache when doing minor changes. The cache can easily forget valuable information about the cached textures by CPU writes or simply by its normal usage.

The original intent of this pull request was to emulate bindless textures at a low level. Instead of handling textures on demand, this new cache would load all textures in the current texture pool and translate from guest texture+sampler indices during shader execution using VK_EXT_descriptor_indexing. This is not possible on OpenGL due to the lack of bindless separate sampler objects, and it can’t be hacked with GL_NV_bindless_texture or GL_ARB_bindless_texture because Nvidia and AMD’s OpenGL driver uses combined samplers for bindless.

A minor issue on emulating this on Vulkan is the lack of heap pointers to image handles on the API (in other words, physical pointers to bindless textures). Games can sample textures with the wrong target and have defined behavior (e.g. sampling 1D as 1D array, and 3D as 1D), we’d have to create one descriptor array per target and fill invalid combinations with null descriptors. Descriptors sets can only have one unbounded array and there are around 8 different texture targets, this wouldn’t fit on some vendors.

Emulating bindless textures this way had to be ditched (for now) for several reasons:
We don’t know what or when textures have been modified from the GPU (imageStore).
Decompressed ASTC textures don’t fit in VRAM on certain titles (2.5+ GiB), although OpenGL drivers can handle this gracefully, even when textures are resident. Compressing them to other formats hurts image quality significantly and it’s slow.
Aliased images have to be hacked or manually synchronized after every write. Because render targets are aligned, it’s sometimes impossible to know if it’s rendering to a 6x8 texture or 7x8, given the report size is 16x8. We can use hints for this (render area, viewport size), but they break when two render targets are bound. Keeping render targets synchronized with aliased textures after every write when bindless is used seems to be the best solution for this.

The idea can still be implemented in the future.

Aliased images are emulated through copies on demand. For example, when a 16x8 texture is rendered and then a texture in the same address is sampled with an 8x8 size, the 16x8 texture is copied to 8x8. This also analyzes mips and layers.

Rendering to compressed textures is implemented with copies. This could be emulated on Vulkan, but it’s not yet implemented due to the constraints required by the API to be able to render compressed images.

On OpenGL, BC4 3D textures are emulated with RGBA8 images (R8 could be used here, but it requires some changes in swizzles that are not implemented). Games rendering to BC4_UNORM textures (The Legend of Zelda: Breath of the Wild) are emulated rendering to the requested RG32UI format and then copied to RGBA8 through compute, decoding it as BC4 in the process. This is currently only known to work on Nvidia’s proprietary driver, the other drivers have bugs that make the copy not work as expected. This has been reduced and reported to mesa’s bug tracker.

Rendering to texture views of a different format is emulated as it should be, drivers that fail to do this properly (Intel and AMD’s proprietary drivers) won’t render properly and there are no intents to fix it on our end. It works as expected on Vulkan on all vendors.

Color blits on Vulkan are emulated with 3D rendering because vkCmdBlitImage lacks format reinterpretations, likely due to an oversight in the API design. Depth stencil blits are not yet implemented on devices that don’t support it natively, and their drivers will assert or ignore this issue.

Thanks to the original design of emulating bindless textures at a low level, this led to design changes that made things faster to emulate. Textures are marked as dirty when written from the CPU, but not removed from the cache. Knowing this, we can cache already visited image views and render targets skipping costly look ups.

Compared to the old cache, there aren’t many new guest features implemented: multisampling, resolves (these are just blits), unnormalized samplers, and sampler textures views with a different lod base (no idea how some games even worked without this).

Code wise, this ditches all virtual function calls from the cache in favor of templated code. All shared pointers were removed, allowing for a simpler life time management. All images, image views, samplers, and framebuffers are stored in a contiguous vector. They are referenced through an index, as pointers would be invalidated on resize.

For this pull request to be mergeable, the following is missing:
- [x] Remove undefined behavior in VideoCommon::SlotVector (it uses reinterpret_cast).
- [x] Add missing documentation
- [x] Move `Common::DivCeil` to a separate pull request.
- [ ] Copies between D32<->R32 and D16<->R16 on OpenGL
- [x] Properly name images and image views on both APIs
- [x] Properly implement clear invalidations (marked commit as hack)
- [x] Insert uniform debug names on both APIs.
- [x] Clean up the code overall.

Credits go to deko3d, this library developed by devkitpro has been crucial for unit testing the cache;
gdkchan from Ryujinx for discussing about different ways to emulate bindless textures, and help me with linked samplers limit;
rip in peri peri for pointing out seamless samplers not being on by default on OpenGL; and
Ryujinx contributors for the code I used as reference for aligning layer sizes and similar functions (pointed out in the code).

I'll squash the commits once everything is ready, keeping them separate for now to ease bisecting.

Fixes #4968, fixes #4750, fixes #3717, fixes #3963, fixes #4979.

### CODE REVIEW

**Comment by ReinUsesLisp** [comment]
Fixes #4968.

File: src/video_core/renderer_opengl/gl_texture_cache.cpp [hunk 1]
```diff
@@ -184,531 +252,860 @@ GLint GetSwizzleSource(SwizzleSource source) {
...
+        name = fmt::format("BufferView {} ({})", NameView(*this), handle);
+    } else {
+        const GLuint parent = image.texture.handle;
+        const GLenum target = ImageTarget(type, image.info.num_samples);
+        glTextureView(handle, target, parent, internal_format, range.base.level,
+                      range.extent.levels, range.base.layer, range.extent.layers);
+        ApplySwizzle(handle, format, info.Swizzle());
+        name = fmt::format("ImageView {} ({})", NameView(*this), handle);
+    }
+    // glObjectLabel(GL_TEXTURE, handle, static_cast<GLsizei>(name.size()), name.data());
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_texture_cache.cpp:996]: If this is left in on purpose there should be a comment indicating that.

Comment (ReinUsesLisp) [reply src/video_core/renderer_opengl/gl_texture_cache.cpp:996]: Sorry, it's part of 157a7f9 , named `HACK: Never name object labels`. I'll address it.

**Comment by choppymaster** [comment]
Fixes  #2771

**Comment by choppymaster** [comment]
Fixes #4750

**Comment by slashiee** [comment]
Fixes #3717 and #3963

**Comment by FearlessTobi** [comment]
https://github.com/yuzu-emu/yuzu/issues/4995 was introduced with this PR.

**Comment by choppymaster** [comment]
Fixes #3262 as a user reported.

**Comment by Aurez** [comment]
I think it should be related to this (sorry and delete if it isn't)

Blade Arcus Rebellion from Shining

1141 : some effects (mostly bloom/smoke) were incorrectly rendered (it was always like that)
[url=https://postimg.cc/XBk6kf12][img]https://i.postimg.cc/XBk6kf12/1141-1.jpg[/img][/url] [url=https://postimg.cc/s1rX7wv4][img]https://i.postimg.cc/s1rX7wv4/1141-2.jpg[/img][/url] [url=https://postimg.cc/YjzS3YFK][img]https://i.postimg.cc/YjzS3YFK/1141-3.jpg[/img][/url]

later versions (until 1156) : the bloom/smoke effects are now correct (congrats!) but there are now many gliches everywere 
[url=https://postimg.cc/LJXFjZtT][img]https://i.postimg.cc/LJXFjZtT/1156-1.jpg[/img][/url] [url=https://postimg.cc/5Q1ZyH98][img]https://i.postimg.cc/5Q1ZyH98/1156-2.jpg[/img][/url] [url=https://postimg.cc/tZ000SLS][img]https://i.postimg.cc/tZ000SLS/1156-3.jpg[/img][/url]

I hope it helps

**Comment by ReinUsesLisp** [comment]
1dd46e7 fixes #4979.

**Comment by ReinUsesLisp** [comment]
@Aurez can you open an issue for that? It seems to be related to blits from the guest OpenGL driver.

**Comment by Aurez** [comment]
> 
> 
> @Aurez can you open an issue for that? It seems to be related to blits from the guest OpenGL driver.

Done : #5009

File: src/video_core/renderer_opengl/gl_texture_cache.cpp (hunk 1 above)
Comment (Nakotei) [reply src/video_core/renderer_opengl/gl_texture_cache.cpp]: Hi, I have the same #4968 issue, but in Animal Crossing 1.5.0. I'm new here and I don't have much programming knowledge, but I want to fix the game. What exactly can I do to fix it?

**Comment by choppymaster** [comment]
I think the below issue is due to tcr (sorry if this is not) 
Mario+Rabbids kingdom battle now have broken graphics. Earlier, this game had perfect graphics but now the graphics is black. Video:

https://youtu.be/0JanMcpHtPU

**Comment by choppymaster** [comment]
I think the below issue is due to tcr (sorry if this is not) 
Mario+Rabbids kingdom battle now have broken graphics. Earlier, this game had perfect graphics but now the graphics is black. Video:

https://youtu.be/0JanMcpHtPU

**Comment by otaku976** [comment]
I have rx 5600xt and have Vulkan bugs in all games. Memory leak too. Can someone give as a clue what is problem and is it possible that all coders work together on gpu fixes. Believe me Vulkan is so buggy, and please forgot for OpenGL for 10 days and all work only on Vulkan and his support for native amd cards. People talking about discrimination about gpu vendor, bcs other emulators work very nice with amd vulkan, many times better than opengl. On the other side we knows that MacOSX using only ati amd gpu chips and all modern console like ps3 ,ps4,ps5, wiiU,  wii, xbox...  On official yuzu page we advised that use latest drivers bcs better VUlkan support, but ppl that is not enough, not fair that we have many problems in graphics and nvidia users havent. You make it better VUlkan and OpenGl support only for nvidia. Please coders give better support for Vulkan on amd gpus.

**Comment by goldenx86** [comment]
@otaku976 OpenGL overs many native advantages over Vulkan, so both will be maintained.
We're working as much as possible to keep stability and feature parity with all vendors, but when emulating Nvidia hardware, you can expect that Geforce products will perform better, not only because of usually better driver support, but also because hardware features and limits are closer to the Nintendo Switch.

**Comment by gidoly** [comment]
i already reported regression on discord but i will write here on document purpose (sorry if it is not needed)
1. xenoblade definitive edition
when first loading area called fallen arm water render white, after re loading area it fix
also have ramleak(more noticeable while play long cutscene)

this is how supposed to look

log:

[xc log.zip](https://github.com/yuzu-emu/yuzu/files/5737520/xc.log.zip)


2.Fate extella/Fate Extella Link
this game got massive ram leak issue after new texture cache arrived on early access
Fate Extella

Fate Extella Link


[fate log.zip](https://github.com/yuzu-emu/yuzu/files/5737521/fate.log.zip)

3. ONINAKI

this game got crash issue when enter dungeun selecting screen



this image is from mainline on new texture cache it crashes when try to load this screen.

[oninaki log.zip](https://github.com/yuzu-emu/yuzu/files/5737551/oninaki.log.zip)

**Review by bunnei — APPROVED** [review]

**Comment by bunnei** [comment]
Lgtm! 💯

**Pull request merged**